core-bundle/tests/Functional/ServiceArgumentsTest.php

```diff
@@ -0,0 +1,184 @@
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Tests\Functional;
+
+use Contao\TestCase\FunctionalTestCase;
+use Psr\Container\ContainerInterface as PsrContainerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Yaml\Tag\TaggedValue;
+use Symfony\Component\Yaml\Yaml;
+
+class ServiceArgumentsTest extends FunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        static::bootKernel();
+    }
+
+    public function testServices(): void
+    {
+        $container = $this->getContainer();
+
+        $files = Finder::create()
+            ->files()
+            ->name('*.yaml')
+            ->name('*.yml')
+            ->path('src/Resources/config')
+            ->in(\dirname(__DIR__, 3))
+        ;
+
+        foreach ($files as $file) {
+            $yaml = Yaml::parseFile($file->getPathname(), Yaml::PARSE_CUSTOM_TAGS);
+
+            if (!isset($yaml['services'])) {
+                continue;
+            }
+
+            foreach ($yaml['services'] as $serviceId => $config) {
+                if ('_' === $serviceId[0]) {
+                    continue;
+                }
+
+                if (!isset($config['class'])) {
+                    continue;
+                }
+
+                if (!$container->has($serviceId)) {
+                    continue;
+                }
+
+                $this->doTestService($serviceId, $config, $container);
+            }
+        }
+    }
+
+    private function doTestService(string $serviceId, array $config, ContainerInterface $container): void
+    {
+        $ref = new \ReflectionClass($config['class']);
+
+        if (!$constructor = $ref->getConstructor()) {
+            $this->markTestSkipped(sprintf('Service %s does not have a constructor method', $serviceId));
+        }
+
+        $arguments = $config['arguments'] ?? [];
+
+        $this->assertGreaterThanOrEqual(
+            $this->countRequiredParameters($constructor),
+            $config['arguments'] ?? [],
+            sprintf('Service %s does not have the necessary amount of parameters.', $serviceId)
+        );
+
+        foreach ($constructor->getParameters() as $i => $parameter) {
+            if (!\array_key_exists($i, $arguments)) {
+                $this->assertTrue($parameter->isOptional(), sprintf('Missing argument %s on service ID "%s".', $i, $serviceId));
+
+                continue;
+            }
+
+            $argument = $arguments[$i];
+            $type = $parameter->getType();
+            $typeName = $type instanceof \ReflectionNamedType ? $type->getName() : null;
+
+            if (null === $argument) {
+                if (!$parameter->allowsNull()) {
+                    $this->addWarning(sprintf('Argument %s ($%s) of %s does not allow NULL, assuming this parameter is set at runtime.', $i, $parameter->getName(), $serviceId));
```

Comment:
Shouldn’t we fail the test here instead of the warning?
And for parameters that we know get set at runtime we explicitly skip this check?

Reply:
how do you _we know_? have a hardcoded list of service argument seems counterinduitive? I actually think we should use valid arguments instead of `~` (e.g. an empty array) for such arguments, but that seems out of scope for this PR.

Reply:
A hardcoded list ist the only way I could think of. The valid cases should never fail or produce a warning (otherwise we’d ignore alle the warnings after a while as it would be impossible to check all of them every time).

Reply:
changed the default values in f30054cb881fe127e3cd267a02eec1c0cb47f856